unified-scheduler-pool/src/lib.rs

```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
     }
 }
 
+pub struct ExecutedTask {
+    task: Task,
+    result_with_timings: ResultWithTimings,
+}
+
+impl ExecutedTask {
+    fn new_boxed(task: Task) -> Box<Self> {
+        Box::new(Self {
+            task,
+            result_with_timings: initialized_result_with_timings(),
+        })
+    }
+}
+
+// A very tiny generic message type to signal about opening and closing of subchannels, which is
+// logically segmented series of Payloads (P1) over a signle continuous time-span, potentially
+// carrying some subchannel metadata (P2) upon opening a new subchannel.
+// Note that the above properties can be upheld only when this is used inside MPSC or SPSC channels
+// (i.e. the consumer side needs to be single threaded). For the multiple consumer cases,
+// ChainedChannel can be used instead.
+enum SubchanneledPayload<P1, P2> {
+    Payload(P1),
+    OpenSubchannel(P2),
+    CloseSubchannel,
+}
+
+type NewTaskPayload = SubchanneledPayload<Task, SchedulingContext>;
+type ExecutedTaskPayload = SubchanneledPayload<Box<ExecutedTask>, ()>;
+
+// A tiny generic message type to synchronize multiple threads everytime some contextual data needs
+// to be switched (ie. SchedulingContext), just using a single communication channel.
+//
+// Usually, there's no way to prevent one of those threads from mixing current and next contexts
+// while processing messages with a multiple-consumer channel. A condvar or other
+// out-of-bound mechanism is needed to notify about switching of contextual data. That's because
+// there's no way to block those threads reliably on such an switching event just with a channel.
+//
+// However, if the number of consumer can be determined, this can be accomplished just over a
+// single channel, which even carries an in-bound control meta-message with the contexts. The trick
+// is that identical meta-messages as many as the number of threads are sent over the channel,
+// along with new channel receivers to be used (hence the name of _chained_). Then, the receiving
+// thread drops the old channel and is now blocked on receiving from the new channel. In this way,
+// this switching can happen exactly once for each thread.
+//
+// Overall, this greatly simplifies the code, reduces CAS/syscall overhead per messaging to the
+// minimum at the cost of a single heap allocation per switching for the sake of Box-ing the Self
+// type to avoid infinite mem::size_of() due to the recursive type structure. Needless to say, such
+// an allocation can be amortized to be negligible.
+enum ChainedChannel<P1, P2> {
```

Comment:
originally, i used condvar, but later i noticed this trick and never looked back. :) hope i explained it well in the above comment...

Reply:
Appreciate the detailed comment here; I think the chained channel idea makes sense. My only real gripe is the visibility to internals within this code, the chained channel is used for synchronization but could easily be screwed up by a change in the code of the handler thread.

It seems possible we can isolate this type and offer an interface for recving that will internally mutate the receiving channel; which will guarantee each receiver can only receive it once (which is currently depending on the handler to overwrite its' receiver).

(pseudo-code with removed generics)
```rust
struct ChainedChannelReceiver<...> {
    receiver: Receiver<...>,
    context: Context,
}

impl ChainedChannelReceiver {
    /// Blocks to receive the next Payload.
    /// If a context-switch is recveived, the context and channel are mutated,
    /// and the next payload message will be blocked on.
    pub fn recv(&mut self) -> Result<Payload> {
        loop {
            match self.receiver.recv()? {
                Payload(p) => return Ok(p),
                PayloadAndChannel((p, new_channel)) => {
                    self.receiver = new_channel;
                    self.context = p;
                }
            }
        }
    }
}

```

Reply:
Not sure my suggestion works if the handler threads will receive from other channels that would actually require a `select!` though.

Reply:
nice suggestion :) 9b5e3383d8e7ed251e5ecba33916ab96c35fb6a5

> Not sure my suggestion works if the handler threads will receive from other channels that would actually require a select! though.

this is good point as well. i did my best in the commit.